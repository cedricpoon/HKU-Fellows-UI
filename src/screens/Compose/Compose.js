import React, { Component } from 'react';
import { Dimensions, Keyboard } from 'react-native';
import { Container } from 'native-base';
import { withNavigation } from 'react-navigation';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Header, ComposeForm } from 'hkufui/components';
import ViewMenu from './ViewMenu/ViewMenu';
import { mapLayoutToState } from 'hkufui/components/helper';
import { localize } from 'hkufui/locale';
import { show3s } from 'hkufui/src/toastHelper';
import { classifyQuery } from 'hkufui/src/screens/Preview/helper';
import NavigationService from 'hkufui/src/NavigationService';

const locale = localize({ country: 'hk', language: 'en' });
const alert = (message) => { show3s({ message, type: 'danger' }); }

export class Compose extends Component {
  constructor(props) {
    super(props);
    this.state = {
      headerLayout: { y: 0, height: 0 },
      title: "",
      subtitle: "",
      hashtags: "",
      content: "",
      native: true
    };
    this._openHeaderMenu = this._openHeaderMenu.bind(this);
    this._renderHeaderMenu = this._renderHeaderMenu.bind(this);
    this._renderNewHeaderMenu = this._renderNewHeaderMenu.bind(this);
    this._renderReplyHeaderMenu = this._renderReplyHeaderMenu.bind(this);
    this._handleTextUpdate = this._handleTextUpdate.bind(this);
    this._toggleNativeMode = this._toggleNativeMode.bind(this);
    this._openReplyHeaderMenu = this._openReplyHeaderMenu.bind(this);
  }

  _openHeaderMenu() {
    const { title, content, hashtags, native } = this.state;
    Keyboard.dismiss();
    // compulsory not filled
    if (title === '' || content === '')
      alert(locale['new.noCompulsory']);
    // not a valid hashtag string
    else if (native && hashtags !== '' && classifyQuery(hashtags) === hashtags)
      alert(locale['new.malformedHashtags']);
    else if (native && this._popup)
      this._popup.toggle();
    else
      // directly preview for Moodle
      NavigationService.navigate('ComposePreview', { title, content, native: false });
  }

  _openReplyHeaderMenu() {
    const { content } = this.state;
    const { navigation } = this.props;
    const { title, subtitle, native } = navigation.state.params;

    if (content === '')
      alert(locale['new.noCompulsory']);
    else if (native && this._popup)
      this._popup.toggle();
    else
      // directly preview for Moodle
      NavigationService.navigate('ComposePreview', { reply: true, title, subtitle, content, native: false });
  }

  _handleTextUpdate(name) {
    return (newValue) => {
      this.setState({ [name]: newValue });
    }
  }

  _toggleNativeMode(isNative) {
    this.setState({ native: isNative });
  }

  _renderHeaderMenu({ width, title, subtitle, hashtags, reply }) {
    const { headerLayout, content } = this.state;

    return (
      <ViewMenu
        position={{ x: width, y: headerLayout.y }}
        parentHeight={headerLayout.height}
        onRef={ref => this._popup = ref}
        payload={{ title, subtitle, hashtags, content, reply }}
      />
    );
  }

  _renderReplyHeaderMenu({ width }) {
    const { navigation } = this.props;

    const nTitle = navigation.getParam('title', null);
    const nSubtitle = navigation.getParam('subtitle', null);

    return this._renderHeaderMenu({ width, title: nTitle, subtitle: nSubtitle, reply: true });
  }

  _renderNewHeaderMenu({ width }) {
    const { title, subtitle, hashtags, native } = this.state;
    return this._renderHeaderMenu({
      width,
      title,
      subtitle: native && subtitle !== '' ? subtitle : null,
      hashtags: native && classifyQuery(hashtags) !== hashtags ? classifyQuery(hashtags) : null,
    });
  }

  render() {
    const { location, navigation } = this.props;
    const { width, height } = Dimensions.get("window");

    const nTitle = navigation.getParam('title', null); // Non null == reply
    const nSubtitle = navigation.getParam('subtitle', null);
    const nNative = navigation.getParam('native', null);

    return (
      <Container>
        { !nTitle ? this._renderNewHeaderMenu({ width }) : this._renderReplyHeaderMenu({ width }) }
        <Header
          title={{ context: !nTitle ? locale['new.header'] : locale['new.replyHeader'] }}
          subtitle={{ context: location }}
          rightIcon='eye'
          onRightPress={nTitle ? this._openReplyHeaderMenu : this._openHeaderMenu}
          backable
          animated={false}
          onLayout={mapLayoutToState('headerLayout', this)}
        />
        <ComposeForm
          onTextUpdates={{
            title: !nTitle ? this._handleTextUpdate('title') : null,
            subtitle: !nTitle ? this._handleTextUpdate('subtitle') : null,
            hashtags: !nTitle ? this._handleTextUpdate('hashtags') : null,
            content: this._handleTextUpdate('content')
          }}
          onToggleMode={!nTitle ? this._toggleNativeMode : null}
          replyParams={!nTitle ? null : {
            title: nTitle,
            subtitle: nSubtitle,
            native: nNative
          }}
          screenHeight={height}
        />
      </Container>
    );
  }
}

Compose.propTypes = {
  location: PropTypes.string,
}

const mapStateToProps = state => ({
  location: state.location.courseTitle
});

const mapDispatchToProps = () => ({

})

export default withNavigation(connect(
  mapStateToProps,
  mapDispatchToProps
)(Compose));
