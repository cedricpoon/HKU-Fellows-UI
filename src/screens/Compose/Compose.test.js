import React from 'react';
import { shallow } from 'enzyme';
import { Compose } from './Compose';

describe('Testing Compose Screen', () => {
  const requiredProps = {
    location: 'location',
    navigation: { getParam: () => {} } // mock of react navigation
  }

  it('renders as expected', () => {
    const wrapper = shallow(
      <Compose {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
