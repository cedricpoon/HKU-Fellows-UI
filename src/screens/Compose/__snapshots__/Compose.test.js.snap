// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Compose Screen renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Compose
    location="location"
    navigation={
      Object {
        "getParam": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <ViewMenu
          onRef={[Function]}
          parentHeight={0}
          payload={
            Object {
              "content": "",
              "hashtags": null,
              "reply": undefined,
              "subtitle": null,
              "title": "",
            }
          }
          position={
            Object {
              "x": 750,
              "y": 0,
            }
          }
        />,
        <Header
          animated={false}
          backable={true}
          onLayout={[Function]}
          onRightPress={[Function]}
          rightIcon="eye"
          rightStyle={
            Object {
              "color": "#4E4A47",
            }
          }
          subtitle={
            Object {
              "context": "location",
            }
          }
          title={
            Object {
              "context": "New Topic",
            }
          }
        />,
        <ComposeForm
          onTextUpdates={
            Object {
              "content": [Function],
              "hashtags": [Function],
              "subtitle": [Function],
              "title": [Function],
            }
          }
          onToggleMode={[Function]}
          replyParams={null}
          screenHeight={1334}
        />,
      ],
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onRef": [Function],
          "parentHeight": 0,
          "payload": Object {
            "content": "",
            "hashtags": null,
            "reply": undefined,
            "subtitle": null,
            "title": "",
          },
          "position": Object {
            "x": 750,
            "y": 0,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animated": false,
          "backable": true,
          "onLayout": [Function],
          "onRightPress": [Function],
          "rightIcon": "eye",
          "rightStyle": Object {
            "color": "#4E4A47",
          },
          "subtitle": Object {
            "context": "location",
          },
          "title": Object {
            "context": "New Topic",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onTextUpdates": Object {
            "content": [Function],
            "hashtags": [Function],
            "subtitle": [Function],
            "title": [Function],
          },
          "onToggleMode": [Function],
          "replyParams": null,
          "screenHeight": 1334,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <ViewMenu
            onRef={[Function]}
            parentHeight={0}
            payload={
              Object {
                "content": "",
                "hashtags": null,
                "reply": undefined,
                "subtitle": null,
                "title": "",
              }
            }
            position={
              Object {
                "x": 750,
                "y": 0,
              }
            }
          />,
          <Header
            animated={false}
            backable={true}
            onLayout={[Function]}
            onRightPress={[Function]}
            rightIcon="eye"
            rightStyle={
              Object {
                "color": "#4E4A47",
              }
            }
            subtitle={
              Object {
                "context": "location",
              }
            }
            title={
              Object {
                "context": "New Topic",
              }
            }
          />,
          <ComposeForm
            onTextUpdates={
              Object {
                "content": [Function],
                "hashtags": [Function],
                "subtitle": [Function],
                "title": [Function],
              }
            }
            onToggleMode={[Function]}
            replyParams={null}
            screenHeight={1334}
          />,
        ],
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onRef": [Function],
            "parentHeight": 0,
            "payload": Object {
              "content": "",
              "hashtags": null,
              "reply": undefined,
              "subtitle": null,
              "title": "",
            },
            "position": Object {
              "x": 750,
              "y": 0,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animated": false,
            "backable": true,
            "onLayout": [Function],
            "onRightPress": [Function],
            "rightIcon": "eye",
            "rightStyle": Object {
              "color": "#4E4A47",
            },
            "subtitle": Object {
              "context": "location",
            },
            "title": Object {
              "context": "New Topic",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onTextUpdates": Object {
              "content": [Function],
              "hashtags": [Function],
              "subtitle": [Function],
              "title": [Function],
            },
            "onToggleMode": [Function],
            "replyParams": null,
            "screenHeight": 1334,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
